exonum/src/runtime/error.rs

```diff
@@ -26,57 +26,112 @@ use crate::{
+///
+/// # Note to Runtime Developers
+///
+/// When should a runtime use different kinds of errors? Here's the guide.
+///
+/// ## `Service` errors
+///
+/// Use `Service` kind if the error has occurred in the service code and it makes sense to notify
+/// users about the error cause and/or its precise kind. These errors are generally raised
+/// if the input data (e.g., the transaction payload) violate certain invariants imposed by the service.
+/// For example, a `Service` error can be raised if the sender of a transfer transaction
+/// in the token service does not have sufficient amount of tokens.
+///
+/// ## `Unchecked` errors
+///
+/// Use `Unchecked` kind if the error has occurred in the service code, and:
+///
+/// - It is caused by the environment (e.g., out-of-memory)
+/// - It should never occur during normal execution (e.g., out-of-bounds indexing, null pointer)
+/// - It otherwise makes no sense to report to users in detail.
+///
+/// This kind of errors generally corresponds to panics in Rust and unchecked exceptions in Java.
+/// `Unchecked` errors are assumed to be reasonably rare by the framework; e.g., they are logged
+/// with a higher priority than other kinds.
+///
+/// Runtime environments can have mechanisms to convert `Unchecked` errors to `Service` ones
+/// (e.g., by catching exceptions in Java or calling [`catch_unwind`] in Rust),
+/// but whether it makes sense heavily depends on the use case.
+///
+/// ## `Dispatcher` errors
+///
+/// Use `Dispatcher` kind if the error has occurred while dispatching the request (i.e., *not*
+/// in the client code). See [`DispatcherError`] for more details.
+///
+/// ## `Runtime` errors
+///
+/// Use `Runtime` kind if the error has occurred in the runtime code and it makes sense to report
+/// the error to the users. A primary example here is artifact deployment: if deployment has failed,
+/// a `Runtime` error can provide more details about the cause.
+///
+/// ## Panics
+///
+/// Panic in the Rust wrapper of the runtime if a fundamental runtime invariant is broken and
+/// continuing node operation is impossible. A panic will not be caught and will lead
+/// to the node termination.
+///
+/// [`catch_unwind`]: https://doc.rust-lang.org/std/panic/fn.catch_unwind.html
+/// [`DispatcherError`]: ../enum.DispatcherError.html
 #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub enum ErrorKind {
-    /// Operation execution has been finished with panic.
-    Panic,
-    /// An error in dispatcher during the execution occurred.
+    /// An unchecked error that has occurred in the service code.
+    ///
+    /// Unlike [`Service`](#variant.Service) errors, unchecked errors do not have a user-defined code.
+    /// Thus, it may be impossible for users to figure out the precise cause of the error;
+    /// they can only use the accompanying error description.
+    Unchecked,
+
+    /// An error in the dispatcher code. For example, the method with the specified ID
```

Comment:
This is not quite correct: i.e. `DispatcherError::UnauthorizedCaller` can occur in the service code and it'll be perfectly fine.